consensus/istanbul/validator/weighted.go

```diff
@@ -586,41 +586,89 @@ func (valSet *weightedCouncil) Refresh(hash common.Hash, blockNum uint64, stakin
 		return errors.New("skip refreshing proposers due to no staking info")
 	}
 
-	// Calculate the Gini coefficient when the following conditions meet.
-	//   1. Need to use the Gini coefficient
-	//   2. Gini coefficient has not yet been calculated
-	//   3. stakingInfo has node info from address book (address book has been activated)
-	if newStakingInfo.UseGini && newStakingInfo.Gini == reward.DefaultGiniCoefficient && len(newStakingInfo.CouncilNodeAddrs) != 0 {
-		calcGiniCoefficientOfValidators(valSet.validators, newStakingInfo)
+	weightedValidators, stakingAmounts, err := valSet.getStakingAmountsOfValidators(newStakingInfo)
+	if err != nil {
+		return err
 	}
+	totalStaking := calcTotalAmount(newStakingInfo, stakingAmounts)
+	calcWeight(weightedValidators, stakingAmounts, totalStaking)
+
+	valSet.refreshProposers(seed, blockNum)
 
-	// Adjust each validator's staking amount by applying the Gini coefficient if necessary.
-	// Also calculate the total staking amount by summing up the staking amount of each validator.
+	logger.Info("Refresh done.", "blockNum", blockNum, "hash", hash, "valSet.blockNum", valSet.blockNum, "stakingInfo.BlockNum", valSet.stakingInfo.BlockNum)
+	logger.Debug("New proposers calculated", "new proposers", valSet.proposers)
+
+	return nil
+}
+
+// getStakingAmountsOfValidators calculates stakingAmounts of validators.
+// If validators have multiple staking contracts, stakingAmounts will be a sum of stakingAmounts with same rewardAddress.
+//  - []*weightedValidator : list of validators which type is converted to weightedValidator
+//  - []float64 : list of stakingAmounts.
+func (valSet *weightedCouncil) getStakingAmountsOfValidators(stakingInfo *reward.StakingInfo) ([]*weightedValidator, []float64, error) {
+	numValidators := len(valSet.validators)
 	weightedValidators := make([]*weightedValidator, numValidators)
 	stakingAmounts := make([]float64, numValidators)
-	totalStaking := float64(0)
+	addedStaking := make([]bool, len(stakingInfo.CouncilNodeAddrs))
+
 	for vIdx, val := range valSet.validators {
 		weightedVal, ok := val.(*weightedValidator)
 		if !ok {
-			return errors.New(fmt.Sprintf("not weightedValidator. val=%s", val.Address().String()))
+			return nil, nil, errors.New(fmt.Sprintf("not weightedValidator. val=%s", val.Address().String()))
 		}
 		weightedValidators[vIdx] = weightedVal
 
-		i, err := newStakingInfo.GetIndexByNodeId(weightedVal.address)
+		sIdx, err := stakingInfo.GetIndexByNodeId(weightedVal.address)
```

Comment:
`weightedVal.address` means the node ID of CCO?
It is set in staking contract?

Reply:
no it isn't. weightedVal.address is a node address of validator in council on platform

Reply:
Then, `GetIndexByNodeId`'s name is not proper.

Reply:
the name of function is changed to GetIndexByNodeAddress